src/sentry/tasks/on_demand_metrics.py

```diff
@@ -310,6 +311,51 @@ def _set_widget_on_demand_state(
         on_demand.save()
 
 
+def set_or_create_on_demand_state(
+    widget_query: DashboardWidgetQuery,
+    organization: Organization,
+    is_low_cardinality: bool,
+    feature_enabled: bool,
+    current_widget_specs,
+):
+    specs = _get_widget_on_demand_specs(widget_query, organization)
+
+    specs_per_version: dict[int, list[HashedMetricSpec]] = {}
+    for hash, spec, spec_version in specs:
+        specs_per_version.setdefault(spec_version.version, [])
+        specs_per_version[spec_version.version].append((hash, spec, spec_version))
+
+    for spec_version in OnDemandMetricSpecVersioning.get_spec_versions():
+        version = spec_version.version
+        specs_for_version = specs_per_version.get(version, [])
+        if not specs:
+            extraction_state = OnDemandExtractionState.DISABLED_NOT_APPLICABLE
+        elif widget_exceeds_max_specs(specs, current_widget_specs, organization):
+            extraction_state = OnDemandExtractionState.DISABLED_SPEC_LIMIT
+        elif not feature_enabled:
+            extraction_state = OnDemandExtractionState.DISABLED_PREROLLOUT
+        elif not is_low_cardinality:
+            extraction_state = OnDemandExtractionState.DISABLED_HIGH_CARDINALITY
```

Comment:
HIGH_CARDINALITY should take precedence over PREROLLOUT since PREROLLOUT is essentially an 'enabled' condition.
```suggestion
        elif not is_low_cardinality:
            extraction_state = OnDemandExtractionState.DISABLED_HIGH_CARDINALITY
        elif not feature_enabled:
            extraction_state = OnDemandExtractionState.DISABLED_PREROLLOUT

```